gh-127794: Validate header name according to rfc-5322

srinivasreddy opened the pull request:
<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
gh-NNNNN: Summary of the changes made
```

Where: gh-NNNNN refers to the GitHub issue number.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `main`.

-->


<!-- gh-issue-number: gh-127794 -->
* Issue: gh-127794
<!-- /gh-issue-number -->

srinivasreddy committed:
gh-127794: Validate header name according rfc-5322 and allow only printable ascii characters

srinivasreddy committed:
Add test case

srinivasreddy committed:
Add some more tests

srinivasreddy committed:
Fix indentation

srinivasreddy committed:
Add news entry

srinivasreddy committed:
Merge branch 'main' into gh-127794

ZeroIntensity commented on Lib/email/message.py:
```diff
@@ -564,6 +564,14 @@ def add_header(self, _name, _value, **_params):
                        filename='Fußballer.ppt'))
         """
+        # Validate header name according to RFC 5322
+        if not _name or ' ' in _name or '\t' in _name or ':' in _name:
```
Could we split this up across multiple lines for readability?

bitdancer commented on Lib/test/test_email/test_email.py:
```diff
@@ -728,6 +728,29 @@ def test_nonascii_add_header_with_tspecial(self):
+            with self.assertRaises(ValueError) as cm:
+               Message().add_header(name, value)
+            self.assertIn(f"Header field name {name!r} contains invalid characters", str(cm.exception))
+
```
Unfortunately this only tests your change to add_header, and does not actually test the underlying problem.  If you implement test methods that test `msg['test header'] = 'foo'`, you will find that the invalid field name will be accepted.  add_header is not used in the `__setitem__` path.

The correct location for the fix is in the two header_store_parse methods in policy.py and _policybase.py.  I'd put a helper method in _policybase and call it from those two methods.  I would also consider using a regex to implement the check...I *think* that will be faster, but you might want to test it.

Also, the tests should test both Message (old API) and EmailMessage (new API).  Since add_header uses the `__setitem__` syntax in its implementation you don't actually need to test it directly, but it wouldn't be bad to keep the tests to make sure that path doesn't get broken in the future.  More tests are good unless they are slow ;)  But not required.

bitdancer replied:
Tests still need to be converted to test `__setitem__` and both `Message` and `EmailMessage`, but I assume you were waiting for that until I approved the code changes.

bitdancer replied:
We have a four way combinatoric here: `__setitem__`, `add_header`, `policy=default`, and `policy=compat32`.  We can get clever and test them all in one nested set of loops:
```
for thispolicy in (policy.default, policy.compat32):
    for method in ('__setitem__', add_header):
        for problem, value in invalid_headers:
            with self.subTest(
                    name=name, 
                    problem=problem,
                    policy=thispolicy.__class__.__name__,
                    method=method,
                    ):
                with self.assertRaises(ValueError) as cm:
                    gatattr( EmailMessage(policy=thispolicy), method(
                        name,
                        problem,
                        )
                ...
```

srinivasreddy committed:
gh-127794: Create a separate function for validating header

srinivasreddy committed:
gh-127794: Coverge all the validations into a single regex

srinivasreddy committed:
gh-127794: Revert the changes

srinivasreddy committed:
gh-127794: Remove one variable

srinivasreddy committed:
gh-127794: Update tests

bitdancer commented on Lib/email/_policybase.py:
```diff
@@ -90,6 +90,14 @@ def __add__(self, other):
         """
         return self.clone(**other.__dict__)
 
+def validate_header(name):
```
s/validate_header/validate_header_name/ for clarity.

bitdancer commented on Lib/email/_policybase.py:
```diff
@@ -90,6 +90,14 @@ def __add__(self, other):
+def validate_header(name):
+    # Validate header name according to RFC 5322
+    import re
+    if not re.match(r'^[^\s:]+$', name):
```
It is better to do the import at the top of the module and compile the regex.  It's a micro optimization, but it is an obvous one.  (Note that utils, which _policybase imports, already imports re, so there's no extra overhead for importing it at the top level here.)

bitdancer commented on Lib/email/_policybase.py:
```diff
@@ -90,6 +90,14 @@ def __add__(self, other):
+        raise ValueError(f"Invalid header field name {name!r}")
+    # Only allow printable ASCII characters
+    if any(ord(c) < 33 or ord(c) > 126 for c in name):
+        raise ValueError(f"Invalid header field name {name!r}")
```
You are using the same error message for both cases here, so there is no advantage to having the tests be separate.  I'd use a variation on the RFC recommended regex for a single test: 
```
header_re = re.compile("[\041-\071\073-\176]+$")
```
Then this whole section becomes:
```
if not header_re.match(name):
    raise ValueError(f"Header field name contains invalid characters: {name!r}")
```

Also, I'd be inclined to put the function def above _PolicyBase, but that's not all that important.

bitdancer commented on Lib/email/policy.py:
```diff
@@ -4,7 +4,7 @@
 import re
 import sys
-from email._policybase import Policy, Compat32, compat32, _extend_docstrings
+from email._policybase import Policy, Compat32, compat32, _extend_docstrings, validate_header
```
Time to wrap this into a multiline expression to keep it under the PEP8 line length limit:
````
from email._policybase import (
    _extend_docstrings,
    Compat32,
    compat32,
    Policy,
    validate_header_name,
    )
```
You'll note that I sorted them alphabetically while I was at it (ignoring case).

bitdancer commented on Lib/test/test_email/test_email.py:
```diff
@@ -728,6 +728,29 @@ def test_nonascii_add_header_with_tspecial(self):
             "attachment; filename*=utf-8''Fu%C3%9Fballer%20%5Bfilename%5D.ppt",
             msg['Content-Disposition'])
 
+    def test_invalid_headers(self):
```
test_invalid_header_names

bitdancer commented on Lib/test/test_email/test_email.py:
```diff
@@ -728,6 +728,29 @@ def test_nonascii_add_header_with_tspecial(self):
+        invalid_headers = [
+            ('Header\x7F', 'Non-ASCII character'),
+            ('Header\x1F', 'control character'),
+        ]
```
What is the reason for having these separate from the list above?  If you did it because you have the checks as two separate clauses in the function, keep in mind that the structure of the tests should be independent of the structure of the code.  It looks like it makes more sense to have them all in one list.

bitdancer commented on Lib/test/test_email/test_email.py:
```diff
@@ -728,6 +728,29 @@ def test_nonascii_add_header_with_tspecial(self):
+            (' LeadingSpace', 'starts with space'),
+            ('TrailingSpace ', 'ends with space'),
+        ]
+        for name, value in invalid_headers:
```
Add: `with self.subTest(name=name, problem=value):` and indent and wrap the remaining lines as appropriate.

bitdancer reviewed (changes requested):
Thanks for the prompt response.  Hopefully I can be reasonably prompt in return, but unfortunately no guarantees :(

srinivasreddy committed:
Address review comments

srinivasreddy committed:
Address review comments

srinivasreddy committed:
Merge branch 'main' into gh-127794

srinivasreddy committed:
Merge branch 'main' into gh-127794

srinivasreddy committed:
Merge branch 'main' into gh-127794

srinivasreddy committed:
Add some more tests

srinivasreddy committed:
Fix the function name

srinivasreddy committed:
Add test case for invalid_header_names

srinivasreddy committed:
Update tests

bitdancer commented on Lib/email/_policybase.py:
```diff
@@ -14,6 +15,15 @@
     'compat32',
     ]
 
+header_re  = re.compile(r'^[^\s:]+$')
```
This re is redundant.  The second re catches spaces as well as non-ascii characters.  Call the second one valid_header_name_re, perhaps.

bitdancer commented on Lib/email/_policybase.py:
```diff
@@ -14,6 +15,15 @@
+def validate_header_name(name):
+    # Validate header name according to RFC 5322
+    if not header_re.match(name):
+        raise ValueError(f"Invalid header field name {name!r}")
```
This if is redundant.  The second if will catch these cases as well.  While I see that you have made the error messages different, there is not IMO enough motivation for doing so.  A whatspace character is still an invalid character, just like a non-ascii or non-printible character is invalid.

bitdancer commented on Lib/test/test_email/test_email.py:
```diff
@@ -728,6 +728,46 @@ def test_nonascii_add_header_with_tspecial(self):
+            with self.subTest(name=name, problem=value):
+                with self.assertRaises(ValueError) as cm:
+                    Message().add_header(name, value)
+                self.assertIn(f"Invalid header field name {name!r}", str(cm.exception))
```
It is better not to so highly constrain the error message.  We really don't care what exactly the error message says.  The important bits are the words 'invalid', 'header', and 'name', and the repr of the header value.  To make the test place only those constraints on the error message, we can write:
```
msg = str(cm.exception)
self.assertRegex(
    msg,
    '(?i)(?=.*invalid)(?=.*header)(?=.*name)',
    )
self.assertIn(f"{name!r}", msg)
````
This uses lookahead assertions to check for each of the substrings in any position in the target string.  Using this as the tests, both versions of your error message pass, as would several other variations that would be equally valid.  And that makes it easy to see that if you delete your first 'if' test in the validator, all the tests still pass.

bitdancer commented on Misc/NEWS.d/next/Library/2024-12-11-17-44-36.gh-issue-127794.VwmRsp.rst:
```diff
@@ -0,0 +1,2 @@
+The :meth:`email.message.Message.add_header` method now validates header
+field names according to :rfc:`5322`.
```
This needs rewriting now, too.  Something like ":meth:email.message.Message now validates header field names according to :rfc`5322` when headers are added via `add_header` or `__setitem__` and raises a ValueError if they contain invalid characters."

srinivasreddy committed:
Remove redundant regular expression

srinivasreddy committed:
Update the regex per review comments

srinivasreddy committed:
Update test cases

srinivasreddy committed:
Improve blurb message

srinivasreddy committed:
Fix failing test

srinivasreddy committed:
Update parameter name

srinivasreddy committed:
Merge branch 'main' into gh-127794

corona10 committed:
gh-115999: Enable BINARY_SUBSCR_GETITEM for free-threaded build (gh-127737)

vstinner committed:
gh-128069: brew link --overwrite tcl-tk@8 to prevent conflict with GitHub image's version (#128090)

brew link --overwrite tcl-tk@8 to prevent conflict with GitHub image's version

Co-authored-by: Hugo van Kemenade <1324225+hugovk@users.noreply.github.com>

kumaraditya303 committed:
gh-128013: fix data race in `PyUnicode_AsUTF8AndSize` on free-threading (#128021)

kumaraditya303 committed:
gh-122706: fix docs for asyncio ssl sockets (#128092)

rruuaanng committed:
gh-127688: Add `SCHED_DEADLINE` and `SCHED_NORMAL` constants to `os` module (GH-127689)

ZeroIntensity committed:
gh-128083: Fix macro redefinition warning in clinic. (GH-127950)

eendebakpt committed:
gh-127951: Add build option to enable pystats on Windows (GH-127952)

colesbury committed:
gh-128008: Add `PyWeakref_IsDead()` (GH-128009)

The `PyWeakref_IsDead()` function tests if a weak reference is dead
without any side effects. Although you can also detect if a weak
reference is dead using `PyWeakref_GetRef()`, that function returns a
strong reference that must be `Py_DECREF()`'d, which can introduce side
effects if the last reference is concurrently dropped (at least in the
free threading build).

markshannon committed:
GH-122548: Implement branch taken and not taken events for sys.monitoring (GH-122564)

nascheme committed:
gh-115999: Specialize `STORE_ATTR` in free-threaded builds. (gh-127838)

* Add `_PyDictKeys_StringLookupSplit` which does locking on dict keys and
  use in place of `_PyDictKeys_StringLookup`.

* Change `_PyObject_TryGetInstanceAttribute` to use that function
  in the case of split keys.

* Add `unicodekeys_lookup_split` helper which allows code sharing
  between `_Py_dict_lookup` and `_PyDictKeys_StringLookupSplit`.

* Fix locking for `STORE_ATTR_INSTANCE_VALUE`.  Create
  `_GUARD_TYPE_VERSION_AND_LOCK` uop so that object stays locked and
  `tp_version_tag` cannot change.

* Pass `tp_version_tag` to `specialize_dict_access()`, ensuring
  the version we store on the cache is the correct one (in case of
  it changing during the specalize analysis).

* Split `analyze_descriptor` into `analyze_descriptor_load` and
  `analyze_descriptor_store` since those don't share much logic.
  Add `descriptor_is_class` helper function.

* In `specialize_dict_access`, double check `_PyObject_GetManagedDict()`
  in case we race and dict was materialized before the lock.

* Avoid borrowed references in `_Py_Specialize_StoreAttr()`.

* Use `specialize()` and `unspecialize()` helpers.

* Add unit tests to ensure specializing happens as expected in FT builds.

* Add unit tests to attempt to trigger data races (useful for running under TSAN).

* Add `has_split_table` function to `_testinternalcapi`.

Xiaokang2022 committed:
gh-128062: Fix the font size and shortcut display of the turtledemo menu (#128063)

Leave the font of the menu bar the default to keep it consistent with the rest of the world. Display the shortcut keys in the right way, using the 'accelerator' option.
---------

Co-authored-by: Peter Bierma <zintensitydev@gmail.com>
Co-authored-by: Terry Jan Reedy <tjreedy@udel.edu>

mpage committed:
gh-127274: Defer nested methods (#128012)

Methods (functions defined in class scope) are likely to be cleaned
up by the GC anyway.

Add a new code flag, `CO_METHOD`, that is set for functions defined
in a class scope. Use that when deciding to defer functions.

tungol committed:
gh-128080: remove unnecessary `__init__` method from Enum (GH-128081)

remove unnecessary __init__ method from Enum

Co-authored-by: Peter Bierma <zintensitydev@gmail.com>

u2rafi committed:
gh-112328: Document EnumDict in docs and release notes (GH-121720)

Co-authored-by: Sviatoslav Sydorenko (Святослав Сидоренко) <wk.cvs.github@sydorenko.org.ua>
Co-authored-by: Ethan Furman <ethan@stoneleaf.us>

vstinner committed:
gh-128058: Fix test_builtin ImmortalTests (#128068)

On 32-bit Free Threading systems, immortal reference count
is 5 << 28, instead of 7 << 28.

Co-authored-by: Peter Bierma <zintensitydev@gmail.com>

hauntsaninja committed:
gh-128030: Avoid error from PyModule_GetFilenameObject for non-module (#128047)

I missed the extra `PyModule_Check` in #127660 because I was looking at
3.12 as the base implementation for import from. This meant that I
missed the `PyModuleCheck` introduced in #112661.

markshannon committed:
GH-122548: Correct magic number comment (GH-128115)

Correct magic number comment

vstinner committed:
gh-109959: Log the current directory in test_glob.test_selflink() (#128122)

vstinner committed:
gh-128116: Skip test_socket VSOCK testStream() on PermissionError (#128120)

ZeroIntensity committed:
gh-127946: Use a critical section for `CFuncPtr` attributes (GH-128109)

encukou committed:
gh-127295: ctypes: Switch field accessors to fixed-width integers (GH-127297)

This should be a pure refactoring, without user-visible behaviour changes.

Before this change, ctypes uses traditional native C types, usually identified 
by [`struct` format characters][struct-chars] when a short (and 
identifier-friendly) name is needed:
- `signed char` (`b`) / `unsigned char` (`B`)
- `short` (`h`) / `unsigned short` (`h`)
- `int` (`i`) / `unsigned int` (`i`)
- `long` (`l`) / `unsigned long` (`l`)
- `long long` (`q`) / `unsigned long long` (`q`)

These map to C99 fixed-width types, which this PR switches to: - 
- `int8_t`/`uint8_t`
- `int16_t`/`uint16_t`
- `int32_t`/`uint32_t`
- `int64_t`/`uint64_t`

The C standard doesn't guarantee that the “traditional” types must map to the 
fixints. But, [`ctypes` currently requires it][swapdefs], so the assumption won't 
break anything.

By “map” I mean that the *size* of the types matches. The *alignment* 
requirements might not. This needs to be kept in mind but is not an issue in 
`ctypes` accessors, which [explicitly handle unaligned memory][memcpy] for the 
integer types.

Note that there are 5 “traditional” C type sizes, but 4 fixed-width ones. Two of 
the former are functionally identical to one another; which ones they are is 
platform-specific (e.g. `int`==`long`==`int32_t`.) This means that one of the 
[current][current-impls-1] [implementations][current-impls-2] is redundant on 
any given platform.


The fixint types are parametrized by the number of bytes/bits, and one bit for 
signedness. This makes it easier to autogenerate code for them or to write 
generic macros (though generic API like 
[`PyLong_AsNativeBytes`][PyLong_AsNativeBytes] is problematic for performance 
reasons -- especially compared to a `memcpy` with compile-time-constant size).

When one has a *different* integer type, determining the corresponding fixint  
means a `sizeof` and signedness check. This is easier and more robust than the 
current implementations (see [`wchar_t`][sizeof-wchar_t] or 
[`_Bool`][sizeof-bool]).


[swapdefs]: https://github.com/python/cpython/blob/v3.13.0/Modules/_ctypes/cfield.c#L420-L444
[struct-chars]: https://docs.python.org/3/library/struct.html#format-characters
[current-impls-1]: https://github.com/python/cpython/blob/v3.13.0/Modules/_ctypes/cfield.c#L470-L653
[current-impls-2]: https://github.com/python/cpython/blob/v3.13.0/Modules/_ctypes/cfield.c#L703-L944
[memcpy]: https://github.com/python/cpython/blob/v3.13.0/Modules/_ctypes/cfield.c#L613
[PyLong_AsNativeBytes]: https://docs.python.org/3/c-api/long.html#c.PyLong_AsNativeBytes
[sizeof-wchar_t]: https://github.com/python/cpython/blob/v3.13.0/Modules/_ctypes/cfield.c#L1547-L1555
[sizeof-bool]: https://github.com/python/cpython/blob/v3.13.0/Modules/_ctypes/cfield.c#L1562-L1572


Co-authored-by: Bénédikt Tran <10796600+picnixz@users.noreply.github.com>

markshannon committed:
GH-127705: Add debug mode for `_PyStackRef`s inspired by HPy debug mode (GH-128121)

shallow-beach committed:
Python Tutorial typo fix (#128077)

bitdancer commented on Lib/email/_policybase.py:
```diff
@@ -14,6 +15,12 @@
+def validate_header_name(name):
+    # Validate header name according to RFC 5322
+    if not valid_header_name_re.match(name):
+        raise ValueError(f"Header field name contains invalid characters: {name!r}")
```
Wrap <80, please

bitdancer commented on Lib/test/test_email/test_message.py:
```diff
@@ -1004,6 +1004,32 @@ def test_folding_with_long_nospace_http_policy_1(self):
+                            method=method,
+                            ):
+                        with self.assertRaises(ValueError) as cm:
+                            getattr(EmailMessage(policy=thispolicy), method)(name, value)
```
Please wrap <80

picnixz replied:
In addition, avoid having multiple calls within the `assertRaises` (you can also use `assertRaisesRegex`). For instance:

```py
message = EmailMessage(policy=thispolicy)
method = getattr(message, method)
pattern = re.escape(...)
with self.assertRaisesRegex(ValueError, pattern) as cm:
	method(name, value)
self.assertIn(repr(name), str(cm.exception))
```

bitdancer commented on Lib/test/test_email/test_message.py:
```diff
@@ -1004,6 +1004,32 @@ def test_folding_with_long_nospace_http_policy_1(self):
+                        with self.assertRaises(ValueError) as cm:
+                            getattr(EmailMessage(policy=thispolicy), method)(name, value)
+                        msg = str(cm.exception)
+                        self.assertRegex( msg, '(?i)(?=.*invalid)(?=.*header)(?=.*name)')
```
Please wrap <80

bitdancer reviewed (changes requested):
Looking good, just some line wrapping needed yet.

Nico-Posada committed:
gh-128049: Fix type confusion bug with the return value of a custom ExceptionGroup split function (#128079)

encukou committed:
gh-112328: Make EnumDict usable on its own and document it (GH-123669)

Co-authored-by: Rafi <rafi.promit@gmail.com>
Co-authored-by: Sviatoslav Sydorenko (Святослав Сидоренко) <wk.cvs.github@sydorenko.org.ua>
Co-authored-by: Ethan Furman <ethan@stoneleaf.us>

picnixz commented on Misc/NEWS.d/next/Library/2024-12-11-17-44-36.gh-issue-127794.VwmRsp.rst:
```diff
@@ -0,0 +1,3 @@
+The :meth:`email.message.Message.add_header` method now validates header
+field names according to :rfc:`5322` when headers are added via add_header
+or __setitem__ and raises a ValueError if they contain invalid characters.
```
```suggestion
The :meth:`email.message.Message.add_header` method now validates header
field names according to :rfc:`5322` and raises a :exc:`ValueError` if
they contain invalid characters. A similar argument applies if headers
are added through :meth:`Message.__setitem__ <object.__setitem__>`.
```

picnixz commented on Lib/email/_policybase.py:
```diff
@@ -14,6 +15,12 @@
     'compat32',
     ]
 
+valid_header_name_re = re.compile("[\041-\071\073-\176]+$")
```
Can't we have a comment saying what those octal values represent? (e.g., "equivalent to pattern "[!-9;-~]+$"). By the way `072` is `:`, so it's still printable by the way (maybe the title of the PR should be changed)

bitdancer replied:
Well, technically "Validate header name according to RFC 5322" is a very precise specification, you just have to refer to the RFC to find out what it is :).  Consequently the PR title *could* be shortened to omit the part after the rfc reference.

The reason to use the numeric version of the regex is because that's how the RFC specifies it.  I have no objection to adding a comment with the ascii version of the regex.

picnixz replied:
> because that's how the RFC specifies

Oh I see (I don't mind keeping the octal values, just maybe a one-line comment would be nice for future readers)

bitdancer replied:
That comment is a tad confusing without any context.  How about `validation regex from RFC, equivalent to ...`

picnixz reviewed (dismissed):
The fact that a ValueError is now raised should be documented in the main documentation file. In addition a What's New entry is needed since it's an incompatible change with 3.13.

So I think we should rather treat this as a feature rather than a bug for now maybe? (if however this is not a feature, we still need a documentation update IMO)

barneygale committed:
GH-127807: pathlib ABCs: remove `PurePathBase._raw_paths` (#127883)

Remove the `PurePathBase` initializer, and make `with_segments()` and
`__str__()` abstract. This allows us to drop the `_raw_paths` attribute,
and also the `Parser.join()` protocol method.

barneygale committed:
GH-127807: pathlib ABCs: remove a few private attributes (#127851)

From `PurePathBase` delete `_globber`, `_stack` and `_pattern_str`, and
from `PathBase` delete `_glob_selector`. This helps avoid an unpleasant
surprise for a users who try to use these names.

barneygale committed:
GH-127807: pathlib ABCs: move private copying methods to dedicated class (#127810)

Move 9 private `PathBase` attributes and methods into a new `CopyWorker`
class. Change `PathBase.copy` from a method to a `CopyWorker` instance.

The methods remain private in the `CopyWorker` class. In future we might
make some/all of them public so that user subclasses of `PathBase` can
customize the copying process (in particular reading/writing of metadata,)
but we'd need to make `PathBase` public first.

gpshead committed:
gh-126664: revert: Use `else` instead of `finally` in docs explaining "with" (#128169)

Revert "gh-126664: Use `else` instead of `finally` in "The with statement" documentation. (GH-126665)"

This reverts commit 25257d61cfccc3b4189f96390a5c4db73fd5302c.

graingert committed:
gh-127949: fix resource warnings in `test_tasks.py` (#128172)

WolframAlph committed:
gh-119786: Fix typos in `InternalDocs/interpreter.md` (#128174)

zanieb committed:
gh-100384: Error on `unguarded-availability` in macOS builds (#128155)

Generate a build error on ``unguarded-availability`` in portable macOS builds (i.e. using MACOSX_DEPLOYMENT_TARGET), preventing invalid use of symbols that are not available in older versions of the OS.

ncoghlan committed:
gh-126180: Remove getopt and optparse deprecation notices (GH-126227)

* Remove getopt and optparse deprecation notices
* Add new docs sections for command line app helper libraries
* Add guidance on choosing a CLI parsing library to the optparse docs
* Link to the new guidance from the argparse and getopt docs
* Reword intro in docs section for superseded stdlib modules
* Reframe the optparse->argparse guide as a migration guide
  rather than as an upgrade guide

---------

Co-authored-by: Serhiy Storchaka <storchaka@gmail.com>

srinivasreddy committed:
Update Misc/NEWS.d/next/Library/2024-12-11-17-44-36.gh-issue-127794.VwmRsp.rst

Co-authored-by: Bénédikt Tran <10796600+picnixz@users.noreply.github.com>

srinivasreddy committed:
Address review comments

srinivasreddy committed:
Merge branch 'main' into gh-127794

srinivasreddy committed:
Update test case

srinivasreddy committed:
Fix test failures

picnixz commented on Lib/email/policy.py:
```diff
@@ -4,7 +4,13 @@
+    compat32,
+    Policy,
+    validate_header_name
+)
```
can you keep a case-sentitive alphabetical order actually? (I think that's how sorted imports elsewhere do it, but check that beforehand)

srinivasreddy replied:
Current looks good.

picnixz replied:
I would still prefer a case-sensitive alphabetical order. We prefer consistency and/or smaller diffs when possible (does "Current looks good." actually tells me that it's how it's done elsewheere?)

srinivasreddy replied:
Adressed in e45165fb873

picnixz commented on Lib/test/test_email/test_email.py:
```diff
@@ -728,6 +728,36 @@ def test_nonascii_add_header_with_tspecial(self):
+                            '(?i)(?=.*invalid)(?=.*header)(?=.*name)'
+                            ) as cm:
+                            class_method(name, value)
+                        self.assertIn(f"{name!r}", str(cm.exception))
```
```suggestion
        for policy in (email.policy.default, email.policy.compat32):
            for setter in (Message.__setitem__, Message.add_header):
                for name, value in invalid_headers:
                    self.do_test_invalid_header_names(policy, setter, name, value)

    def do_test_invalid_header_names(self, policy, setter, name, value):
        with self.subTest(policy=policy, setter=setter, name=name, value=value):
            message = Message(policy=policy)
            pattern = r'(?i)(?=.*invalid)(?=.*header)(?=.*name)'
            with self.assertRaisesRegex(ValueError, pattern) as cm:
                 setter(message, name, value)
            self.assertIn(f"{name!r}", str(cm.exception))
```

This could reduce the number of lines.

srinivasreddy replied:
I am not sure about this optimisation. I would always have a full context instead of splitting into another function that reduces lines.

srinivasreddy replied:
What do you think @bitdancer ?

picnixz replied:
> I am not sure about this optimisation. I would always have a full context instead of splitting into another function that reduces lines.

It's for readability and to ease the addition of more cases (also, please do not mark the discussion as resolved until it is really (you asked for bitdancer's opinion without waiting for them to reply)) (and for me to check).

FTR: if you want to acknowledge the suggestion, you can also comment with "Done in COMMIT_SHA".

picnixz commented on Lib/test/test_email/test_message.py:
```diff
@@ -1004,6 +1004,35 @@ def test_folding_with_long_nospace_http_policy_1(self):
+            ('Header\x7F', 'Non-ASCII character'),
+            ('Header\x80', 'Extended ASCII'),
+        ]
+        for thispolicy in (policy.default, policy.compat32):
```
Ditto for this test (IMO it would be a bit clearer)

picnixz commented on Lib/email/policy.py:
```diff
@@ -140,6 +146,7 @@ def header_store_parse(self, name, value):
         """
         if hasattr(value, 'name') and value.name.lower() == name.lower():
             return (name, value)
+        validate_header_name(name)
```
Should we validate the header name if the value has a "name"?

srinivasreddy replied:
Do you mean it would be better to change the function name to `validate_header(name)` ?

picnixz replied:
No, I meant: should we also call `validate_header_name` in the `if hasattr(value, 'name') and value.name.lower() == name.lower():` block before returning `(name, value)` as is? namely:

```py
if hasattr(value, 'name') and value.name.lower() == name.lower():
    validate_header_name(name)
    return (name, value)
validate_header_name(name)
```

srinivasreddy replied:
Addressed in 2550011a71e. I do not remember the reason for placing `validate_header_name` between two if blocks. Make sense to move it up instead of calling twice?

picnixz replied:
Totally. I just wanted to be verbose but yes it just makes sense to call it directly.

srinivasreddy committed:
Update Lib/test/test_email/test_email.py

Co-authored-by: Bénédikt Tran <10796600+picnixz@users.noreply.github.com>

srinivasreddy committed:
Address review comments

srinivasreddy committed:
Merge branch 'main' into gh-127794

picnixz commented on Lib/test/test_email/test_email.py:
```diff
@@ -728,6 +728,31 @@ def test_nonascii_add_header_with_tspecial(self):
+            with self.assertRaisesRegex(ValueError, pattern) as cm:
+                 setter(message, name, value)
+            self.assertIn(f"{name!r}", str(cm.exception))
+
```
```suggestion
```

srinivasreddy replied:
addressed in 8233e928a8d

srinivasreddy committed:
Order imports as per  case sensitive alphabetical order

srinivasreddy committed:
Remove the extraneous line

srinivasreddy committed:
Move the validate_header_name up

srinivasreddy committed:
Merge branch 'main' into gh-127794

picnixz commented on Lib/email/_policybase.py:
```diff
@@ -14,6 +15,14 @@
+# equivalent to pattern re.compile("[!-9;-~]+$")
+valid_header_name_re = re.compile("[\041-\071\073-\176]+$")
+
+def validate_header_name(name):
```
Actually, I just observed that we also have `email.utils`. I don't know whether this one is better to be placed here or in the utils, next to `_has_surrogates`. @bitdancer any thoughts? Also, we seem to use underscore names for private utils even if the module is already private so maybe `_validate_header_name` can be used as well (this is just a suggestion, but I'd like to hear others' thoughts)

bitdancer replied:
Well, the function has no leading underscore because it is a "public" function of the `_policybase` module; that is, it is part of the `_policybase` API and won't change.  But it is not a *public* function because `_policybase` is not public/documented.  That said, it wouldn't be crazy to put it in utils and document it.  But let's not do that absent an actual use case.  `_has_surrorages` is in utils because lots of other modules use it.  `validate_header_name` should, in theory, be used only by modules derived from `_policybase`, since the policy is what determines whether a header name is valid.  I'm inclined to leave it here absent a use case, and since it is private it can be more easily moved later if there  is a need.

picnixz commented on Misc/NEWS.d/next/Library/2024-12-11-17-44-36.gh-issue-127794.VwmRsp.rst:
```diff
@@ -0,0 +1,4 @@
+The :meth:`email.message.Message.add_header` method now validates header
+field names according to :rfc:`5322` and raises a :exc:`ValueError` if
```
Can we have the dedicated section with field names specifications? If you find it, you can link it as <code>:rfc:\`5322#section-X.Y\`</code> and this would users knowing about it instead of going through the entire RFC.

srinivasreddy replied:
Addressed in 56a77e93524

picnixz reviewed (commented):
Some final thoughts. Otherwise, it looks fine to me.

srinivasreddy committed:
Add RFC section

picnixz commented on Misc/NEWS.d/next/Library/2024-12-11-17-44-36.gh-issue-127794.VwmRsp.rst:
```diff
@@ -0,0 +1,5 @@
+The :meth:`email.message.Message.add_header` method now validates header
+field names according to :rfc:`RFC 5322 section 3.6.8 <5322#section-3.6.8>`
```
```suggestion
field names according to :rfc:`RFC 5322, Section 3.6.8 <5322#section-3.6.8>`
```

How does

```rst
:rfc:`5322#section-3.6.8`
```

renders? it should be picked up automatically by Sphinx but I don't remember how it's being rendered. If it's rendered fine, we can keep a shorter formulation.

picnixz reviewed (commented):
I forgot to post my review 3 days ago..

srinivasreddy committed:
Update Misc/NEWS.d/next/Library/2024-12-11-17-44-36.gh-issue-127794.VwmRsp.rst

Co-authored-by: Bénédikt Tran <10796600+picnixz@users.noreply.github.com>

picnixz reviewed (approved):
Once RTD is back online, we can merge this one I think.

srinivasreddy committed:
Merge branch 'main' into gh-127794

srinivasreddy committed:
Merge branch 'main' into gh-127794

picnixz reviewed (commented):
Friendly ping @bitdancer

I still don't know whether to treat it as a bug or a feature. I would say a bugfix because before it didn't work at all. However, we are now raising an exception; yet the code that previously didn't work also didn't raise (at least, according to the issue description). We can say that it's an example of "garbage in -> garbage out" (even though the garbage out is not the expected garbage out). Now that we raise, existing (but faulty) code could actually be raising exceptions which could not be handled by an application. And this can be considered a breaking change.

Therefore, should we backport this? 3.12 is nearing it's EOL in terms of bugfix so maybe it's a bit late to backport to it? I don't really know the policy in this case. cc @encukou @serhiy-storchaka  @vstinner as they know better what to do in this situation.

bitdancer commented on Misc/NEWS.d/next/Library/2024-12-11-17-44-36.gh-issue-127794.VwmRsp.rst:
```diff
@@ -0,0 +1,5 @@
+field names according to :rfc:`RFC 5322, Section 3.6.8 <5322#section-3.6.8>`
+and raises a :exc:`ValueError` if they contain invalid characters. A similar
+argument applies if headers are added through :meth:`Message.__setitem__
+<object.__setitem__>`.
```
I would phrase this as follows:

```
When headers are added to :meth:`email.message.Message` objects, either through
:meth:`email.message.Message.__setitem__` or :meth:`email.message.Message.add_header`,
the field name is now validated according to  :rfc:`RFC 5322, Section 3.6.8 <5322#section-3.6.8>`
and a :exc:`ValueError` is raised if the field name contains any invalid characters.
```

bitdancer commented on Lib/test/test_email/test_email.py:
```diff
@@ -728,6 +728,30 @@ def test_nonascii_add_header_with_tspecial(self):
+        for policy in (email.policy.default, email.policy.compat32):
+            for setter in (Message.__setitem__, Message.add_header):
+                for name, value in invalid_headers:
+                    self.do_test_invalid_header_names(policy, setter, name, value)
```
Technically this should be wrapped <80.

bitdancer commented on Lib/test/test_email/test_message.py:
```diff
@@ -1004,6 +1004,30 @@ def test_folding_with_long_nospace_http_policy_1(self):
+            with self.assertRaisesRegex(ValueError, pattern) as cm:
+                 setter(message, name, value)
+            self.assertIn(f"{name!r}", str(cm.exception))
+
```
Duplicated test?  Did you forget to delete it from test_email?  test_message does seem to be the better location for it.

srinivasreddy replied:
I was asked to write tests here as well - Please refer https://github.com/python/cpython/pull/127820#discussion_r1884632081

bitdancer replied:
Ah, right, I forgot that I left Message tests in test_email and only put EmailMessage tests in test_message.  So what you did is correct.

bitdancer reviewed (changes requested):
Sorry it took me so long to get back to this :(

srinivasreddy committed:
Address review comments

srinivasreddy committed:
Merge branch 'main' into gh-127794

bitdancer reviewed (approved):
LGTM

bitdancer committed:
Merge branch 'main' into gh-127794

picnixz commented on Lib/email/_policybase.py:
```diff
@@ -14,6 +15,14 @@
     'compat32',
     ]
 
+# validation regex from RFC 5322,equivalent to pattern re.compile("[!-9;-~]+$")
```
```suggestion
# validation regex from RFC 5322, equivalent to pattern re.compile("[!-9;-~]+$")
```

picnixz committed:
fixup typo

picnixz commented on Misc/NEWS.d/next/Library/2024-12-11-17-44-36.gh-issue-127794.VwmRsp.rst:
```diff
@@ -0,0 +1,4 @@
+When headers are added to :meth:`email.message.Message` objects, either through
```
```suggestion
When headers are added to :class:`email.message.Message` objects, either through
```

picnixz committed:
fixup rst

picnixz commented on Misc/NEWS.d/next/Library/2024-12-11-17-44-36.gh-issue-127794.VwmRsp.rst:
```diff
@@ -0,0 +1,4 @@
+When headers are added to :class:`email.message.Message` objects, either through
+:meth:`email.message.Message.__setitem__` or :meth:`email.message.Message.add_header`,
+the field name is now validated according to  :rfc:`RFC 5322, Section 3.6.8 <5322#section-3.6.8>`
+and a :exc:`ValueError` is raised if the field name contains any invalid characters.
```
```suggestion
When headers are added to :class:`email.message.Message` objects, either through
:meth:`email.message.Message.__setitem__` or :meth:`email.message.Message.add_header`,
the field name is now validated according to :rfc:`RFC 5322, Section 2.2 <5322#section-2.2>`
and a :exc:`ValueError` is raised if the field name contains any invalid characters.
```

Actually, the definition of a header field name is in §2.2. 3.6.8 was only for optional fields but the requirements are identical for the name (2.2 says "all characters from 33 to 126 except colon and 3.6.8 says all printable characters except space and colon, but space is 32 so it's the same).

picnixz committed:
Update Misc/NEWS.d/next/Library/2024-12-11-17-44-36.gh-issue-127794.VwmRsp.rst
